Fixes and improvements

Mattijah opened the pull request:
Various fixes and improvements. Some of the more important ones improve performance by fixing blocking of the main thread.

JavaNoobsyBoobsy commented:
These are valuable changes. Thank you very much for your great contribution.

JeromeCHA commented on Merchant/App/Scanner/ScannerViewController.swift:
```diff
@@ -22,19 +22,13 @@ class ScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDel
-        view.backgroundColor = UIColor.white
-        
-        checkPermission()
+        view.backgroundColor = .black
```
@Mattijah **[QUESTION]**
Why did you change the background to black? Moreover, is there a reason why you removed the `checkPermission()` here?

Mattijah replied:
**Black background**
- As described in the commit message this change reduces the initial "flash" when switching from the white background to the actual camera view (which is always darker). I'm not a designer, but I do feel like it's more natural like this. I would also suggest to present this view controller modally and not pushing it into a navigation stack. 

**checkPermission**
- This is part of the e5d07a2c776ee692bb15cdb610c1ae64616dd6cb and explained here https://github.com/Bitcoin-com/bitcoin-cash-merchant-ios/pull/3/commits/e5d07a2c776ee692bb15cdb610c1ae64616dd6cb#r290988431

JeromeCHA commented on Merchant/App/Scanner/ScannerViewController.swift:
```diff
@@ -135,9 +131,9 @@ class ScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDel
-        setupCamera()
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        checkPermission()
```
@Mattijah **[QUESTION]**
Is there any reason why did you remove the setup camera here?

Mattijah replied:
Copied: _"It is not removed just moved into viewDidLoad. This method also calls setupCamera which is being called in the viewDidLoad as well, so essentially you call setupCamera twice."_

JeromeCHA commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -220,27 +220,21 @@ extension SettingsViewController: UITableViewDelegate, UITableViewDataSource {
         return textField.resignFirstResponder()
     }
-
+    
```
@Mattijah **[MUST]**
Unneeded space. Please remove it

Mattijah replied:
This offset was (is) added automatically by Xcode for every blank line. I can remove it but I feel like there are many more lines that have or don't have this. I think it would be nice to fix them all in 1 commit across the entire project and add the preferred settings to some kind of a swift-style-guide.

jbdtky commented on Merchant/App/PaymentInput/PaymentInputViewController.swift:
```diff
@@ -11,26 +11,26 @@ import BDCKit
 class PaymentInputViewController: PinViewController {
     
-    var amountLabel: BDCLabel = {
+    let amountLabel: BDCLabel = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possible many more places than just these...

jbdtky replied:
Yes, but you change it from `var` to `let`, please add the `fileprivate` in the same time.

Mattijah replied:
Well, as I mentioned this commit was only about correcting the value types not the access level. If I was going to correct the access level I would need to touch other places too, so essentially I would be doing 2 different things in one commit. If you would ask me then to revert either of them (type or the access level) that would not be so easy. That's why I haven't done it all in 1 time. I try to isolate the changes. 🙂

jbdtky replied:
You spent time to check every single variable to change it as a constant. So I am sure you have time to change the access level 😄

Mattijah replied:
- I didn't check every single variable I just simply used the Xcode search which filtered that for me 😉
- It's not about having or not having time. It was about not doing 5 different things in 1 commit.

jbdtky commented on Merchant/App/PaymentInput/PaymentInputViewController.swift:
```diff
@@ -11,26 +11,26 @@ import BDCKit
         return label
     }()
-    var amountView: UIView = {
+    let amountView: UIView = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possible many more places than just these...

jbdtky commented on Merchant/App/PaymentInput/PaymentInputViewController.swift:
```diff
@@ -11,26 +11,26 @@ import BDCKit
     }()
     
-    var alertStackView: UIStackView = {
+    let alertStackView: UIStackView = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possible many more places than just these...

jbdtky commented on Merchant/App/PaymentInput/PaymentInputViewController.swift:
```diff
@@ -11,26 +11,26 @@ import BDCKit
     }()
     
-    var blurView: UIVisualEffectView = {
+    let blurView: UIVisualEffectView = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

jbdtky commented on Merchant/App/Pin/PinController.swift:
```diff
@@ -11,7 +11,7 @@ import BDCKit
 class PinController: PinViewController {
 
-    var pinMessageLabel: BDCLabel = {
+    let pinMessageLabel: BDCLabel = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

jbdtky commented on Merchant/App/Pin/PinController.swift:
```diff
@@ -21,7 +21,7 @@ class PinController: PinViewController {
     }()
 
-    var pinView: UIView = {
+    let pinView: UIView = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

JeromeCHA commented on Merchant/Core/Services/SocketService.swift:
```diff
@@ -53,15 +54,11 @@ class SocketService {
             
-            guard let transaction = try? JSONDecoder().decode(WebSocketTransactionResponse.self, from: data!) else {
-                return
+            if let transaction = try? JSONDecoder().decode(WebSocketTransactionResponse.self, from: data!) {
```
@Mattijah **[IMO]**
I know it was in the previous code too, but here data should not be forced unwrap. Please use guard instead.

jbdtky replied:
Yes, well found.

Mattijah replied:
I agree with you guys, but again this commit was about something else. I didn't really wanna fix forced unwrapping as a part of commit that deals with something else. If I was then ask to revert this commit (which changes guards into ifs) I would then also undo the fix **or** I would need to go and revert all changes manually. It doesn't make any sense. I hope you understand me.

jbdtky replied:
It’s a PR. We are not here to point fingers on who did what but to improve the project. He pointed out you an issue. I think this problem as well as fixing the leak that I wrote with the web socket are more relevant than changing every single var to let or removing comment.

Mattijah replied:
Sorry, I thought "_Please use guard instead_" was directed to me.

> I think removing a force wrap as well as fixing the leak that I wrote with the web socket are more relevant than changing every single var to let or removing comment.

- The leak is fixed
- There were probably 20x more properties with incorrect type than force unwrap. I simply noticed it first and correct it. Apart from that I did another 12 changes I can't fix it all. There is still plenty of stuff left.

jbdtky commented on Merchant/App/Scanner/ScannerViewController.swift:
```diff
@@ -135,9 +131,9 @@ class ScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDel
-        setupCamera()
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        checkPermission()
```
Cf. Life cycle of a view controller. ViewDidLoad will be triggered without the ViewDidAppear. No change required.

Mattijah replied:
Could you please elaborate more on this. I'm not sure what do you mean. `viewDidLoad` > `viewWillAppear` > `viewDidAppear`.

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
     fileprivate let cellId = "settingsCell"
     
-    var tableView: UITableView = {
+    let tableView: UITableView = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
-    var selectedCurrencyLabel = BDCLabel.build(.subtitle)
-    var currenciesPickerView = UIPickerView(frame: .zero)
-    var currenciesView: UIView = {
+    let companyNameTextField = BDCTextField.build(.type1)
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
-    var currenciesPickerView = UIPickerView(frame: .zero)
-    var currenciesView: UIView = {
+    let companyNameTextField = BDCTextField.build(.type1)
+    let destinationAddressTextField = BDCTextField.build(.type1)
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
-    var currenciesView: UIView = {
+    let companyNameTextField = BDCTextField.build(.type1)
+    let destinationAddressTextField = BDCTextField.build(.type1)
+    let pinCodeLabel = BDCLabel.build(.subtitle)
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possibly many more places than just these...

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
+    let companyNameTextField = BDCTextField.build(.type1)
+    let destinationAddressTextField = BDCTextField.build(.type1)
+    let pinCodeLabel = BDCLabel.build(.subtitle)
+    let selectedCurrencyLabel = BDCLabel.build(.subtitle)
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possible many more places than just these...

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
+    let destinationAddressTextField = BDCTextField.build(.type1)
+    let pinCodeLabel = BDCLabel.build(.subtitle)
+    let selectedCurrencyLabel = BDCLabel.build(.subtitle)
+    let currenciesPickerView = UIPickerView(frame: .zero)
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possible many more places than just these...

jbdtky commented on Merchant/App/Settings/SettingsViewController.swift:
```diff
@@ -14,18 +14,18 @@ class SettingsViewController: BDCViewController {
+    let pinCodeLabel = BDCLabel.build(.subtitle)
+    let selectedCurrencyLabel = BDCLabel.build(.subtitle)
+    let currenciesPickerView = UIPickerView(frame: .zero)
+    let currenciesView: UIView = {
```
fileprivate please

Mattijah replied:
In this commit was the objective to correct the value type only. Talking about the access control I would leave this for another commit which will do only that as there are possible many more places than just these...

jbdtky commented on Merchant/Core/Services/SecureAccessService.swift:
```diff
@@ -19,9 +19,7 @@ class SecureAccessService: PinDelegate {
-        guard let window = UIApplication.shared.keyWindow
-            , let currentViewController = window.rootViewController else {
-                // Manage the error here
+        guard let window = UIApplication.shared.keyWindow, let currentViewController = window.rootViewController else {
```
No change required.

Mattijah replied:
This is part of the commit which removed unnecessary comments I believe. Here was a comment saying `Manage the error here`. If there is an error to manage it should be managed and not just a comment saying that...

jbdtky replied:
You can change the comment by `TODO :`. However I was not talking about the comment but the code style.

Mattijah replied:
Well, you haven't really said that 😅Talking about `TODO:` you are completely right and I've been thinking about changing some of the comments to `TODO`. The reason I didn't is because I haven't found any comments like that and because if there is anything "TO-DO" I think it should be simply done/scheduled and not pushed into a production.

Talking about style of the code. Again, I don't think you guys have here any style-guide that I could follow. If I remember correctly I also think there can be found both alternatives in the project.

jbdtky commented on Merchant/Core/Services/SocketService.swift:
```diff
@@ -36,10 +36,11 @@ class SocketService {
     
-    init() {
+    private init() {
         self.ws.event.close = { code, reason, clean in
```
weak required.

Mattijah replied:
What do you mean? There is a `weak` specified, but wasn't before.

jbdtky commented:
> @Mattijah Thank you for tour PR. We appreciate that.
> However, I have just commented it. Please check it. 🙂
> Also, why did you removed almost all of the comments?

Yes, please reverse the comments please.

Mattijah commented on Merchant/App/Scanner/ScannerViewController.swift:
```diff
@@ -22,19 +22,13 @@ class ScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDel
-        
-        view.backgroundColor = UIColor.white
-        
-        checkPermission()
```
It is not removed just moved into `viewDidLoad`. This method also calls `setupCamera` which is being called in the `viewDidLoad` as well, so essentially you call `setupCamera` twice.

jbdtky replied:
Yes, correct. However, do you think it is optimal to call it in the `viewWillAppear`?

Mattijah replied:
What is the advantage of calling it inside the `viewDidLoad`? (bear in mind it also initialises the camera).

Mattijah commented on Merchant/App/Scanner/ScannerViewController.swift:
```diff
@@ -22,19 +22,13 @@ class ScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDel
-        view.backgroundColor = UIColor.white
-        
-        checkPermission()
+        view.backgroundColor = .black
```
As described in the commit message this change reduces the initial "flash" when switching from the white background to the actual camera view (which is always darker). I'm not a designer, but I do feel like it's more natural like this. I would also suggest to present this view controller modally and not pushing it into a navigation stack.

jbdtky replied:
It is a design decision. Thanks for you proposal how we will keep the existing color.

Mattijah replied:
Do you have an actual designer working on it? I'm asking this question because the design itself looks quite outdated. Also to give you an example of other apps which implement the same style so black initial colour for the camera view: iOS built-in apps, messenger, instagram... I haven't really seen white one yet.

Nevertheless, if the white one is the one you want I will absolutely respect that. If you would like me to revert this commit please confirm this to me. Thanks you 😊

jbdtky replied:
Yes, please.

Mattijah replied:
Playing a little bit more with the app I decided to go and remove the reverting commit back. The reason I did that is simply because you are not right. 

- You have probably not even looked at it. Haven't seen the difference for yourself.
- You are refusing to have any constructive debate, no design based arguments and instead just insisting on reverting the change back. Like a little child.
- You are not taking into consideration the fact that literally all apps available use the same setting. We are talking about apps created by gigantic companies with professional designers behind them. Are you saying they are not right? The white background colour is simply not correct. It is not correct as many other UI elements or even pieces of the code in this project.

Just to be clear, I'm not interested in fighting and arguing here. I'm interested in constructive debates and right decision making. Unless, you have any arguments and can explain why the white colour is better I'm sorry, but I will not revert it back.

jbdtky commented on Merchant/App/Scanner/ScannerViewController.swift:
```diff
@@ -135,9 +131,9 @@ class ScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDel
-        setupCamera()
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        checkPermission()
```
I didn't see your other change.

Mattijah commented:
Guys do you really need comments saying:

- "Date" above a "DateFormatter" https://github.com/Bitcoin-com/bitcoin-cash-merchant-ios/pull/3/commits/30b96857c6f21132dce92dd1bbfa4e82dbead9eb#diff-9ed4e12c68b26427e4dde284f00aea20L42

- Or saying that the string has to be at least 8 chars long just above an if statement `satoshisStr.count < 9` https://github.com/Bitcoin-com/bitcoin-cash-merchant-ios/pull/3/commits/30b96857c6f21132dce92dd1bbfa4e82dbead9eb#diff-de1afc1a07780d8037d7e9c460375d59L24

and many others. No offence but these comments are usually left by junior developers. If you insist on reverting this change I will but honestly... the code should be (and in many cases is) clear enough.

jbdtky commented:
I think some comments as you mentioned are not relevant, however others can be.

Mattijah commented:
@jbdtky can you give me please an example? I will then go and revert them individually.

Mattijah commented:
@jbdtky @JeromeCHA alright, I think I addressed all the comments. If there is anything else I can do or if I missed something please just let me know. I hope I haven't made you guys too mad (especially not the author of this project 😂). That definitely wasn't my intention.

magmahindenburg commented:
Thank you @Mattijah for your valuable and good contribution to our open source merchant app. This is a big improvement. We truly appreciate your efforts.
